Set correct ghost layer for multigrid levels with periodic boundary conditions

kronbichler opened the pull request:
The previous implementation of the multigird levels in parallel::distributed::Triangulation did not take periodic boundary conditions into account. This only worked for the active cells through the information provided by p4est, but not for the multigrid levels that we construct manually based on the vertices around the owned level sets. For periodic boundary conditions, we need to look at both vertices of a periodic face and make sure they are set together. To this end, I added a new method in parallel::distributed::Triangulation that is used on the assignment of ghost cells in the triangulation and the distribution of degrees of freedom (two places in the parallel distributed dof handler policy). To make this work, the periodic face pairs that are passed to the triangulation by `add_periodicity()` need to be stored.

The method is only tested for uniform meshes (since parallel multigrid is not working with adaptivity yet) and standard orientations over periodic faces. The test that comes with this PR only outputs degrees of freedom and possible ghost information to make it a 'unit test', but the correctness of the implementation has been asserted with a full multigrid solver for DG on grids with standard orientation. (And the most difficult thing has been to make the code go past the various communication steps where the code could lock due to two neighboring processors disagreeing on what ghosts are to be present, so the test really covers the aspects of this PR.)

Checked that all `multigrid` and `mpi` tests are still working.

tjhei commented:
Nice. I have to look a little closer, but I have a couple of questions already:
- add_periodicity can be called more than once and adds the new periodicity to the existing ones. I am not sure we ever tried/used this, but your code assumes this is not the case.
- You keep mentioning the faces have two vertices: what about the 3d case?

tjhei commented:
I try to make the distinction between "ghost on lvl i" and "level ghost on lvl i" and I would like to ask you make that distinction a little bit more clear in a few places. "level ghost" implies looking at level_subdomain_id and all neighbors are on the same level (so there is no level ghost neighbor if the neighbor is coarser even if there is an active ghost neighbor).

Would it make sense to separate the level functionality of ```fill_vertices_with_ghost_neighbors``` into a separate function to help with this distinction? It looks like there is no overlap in the implementation anyways. I admit the old naming ```mark_locally_active_vertices_on_level``` is pretty bad.

kronbichler commented:
> add_periodicity can be called more than once and adds the new periodicity to the existing ones. I am not sure we ever tried/used this, but your code assumes this is not the case.

That's a good point. I will need to call push_back to the vector of face pairs rather than calling operator=. (We could even have the same face pair twice.) The only thing I need to think about a bit more is whether the iterators in the periodic face pairs remain valid. They should because adding periodicity does not change the way the mesh is passed through even though a new mesh is created internally.

Since we need to clear a triangulation when starting over anyway, push_back rather than equal sign should be fine.

kronbichler commented:
> You keep mentioning the faces have two vertices: what about the 3d case?

The 'two vertices' really refers to the periodicity where two vertices refer to the same 'degree of freedom'. It is not referring to the fact that a vertex will belong to up to four faces in 3D. I will check the documentation whether I can be more clear.

kronbichler commented:
> Would it make sense to separate the level functionality of fill_vertices_with_ghost_neighbors into a separate function to help with this distinction?

So you suggest to put the 'level' part of the `fill_vertices_with_ghost_neighbors` into a function `fill_level_vertices_with_ghost_neighbors`? You are right that the 'level' implementation is disjoint from the active part (that way, the old 'active' code would remain untouched, whereas I now had to change the indentation level). This sounds like a good idea.

@tjhei do you have other suggestions regarding your second comment?

tjhei commented:
> The 'two vertices' really refers to the periodicity where two vertices refer to the same 'degree of freedom'. It is not referring to the fact that a vertex will belong to up to four faces in 3D. 

Doh, of course. My bad. :-)

> So you suggest to put the 'level' part of the fill_vertices_with_ghost_neighbors into a function fill_level_vertices_with_ghost_neighbors?

Yes.

> do you have other suggestions regarding your second comment?

I was also thinking of the documentation of the function ```fill_vertices_with_ghost_neighbors```. It wasn't clear to me if it operated on "ghosts" or "level ghosts" if you supply the level. By splitting the functions this is a lot more obvious anyways.

One more thing:

> The method is only tested for uniform meshes

Distributing MG dofs should already work with adaptive refinement. Did you check if this works with periodic faces? If yes we should probably add a test.

kronbichler commented:
@tjhei I put the `fill_level_vertices_with_ghost_neighbors` into a separate functions, fixed calling `add_periodicity` several times and added tests for those situations (`mg_ghost_dofs_periodic_02`). Moreover, I now also test that the clear() method actually cleans up everything (more important now that we append data to the vector of periodic face pairs), `mg_ghost_dofs_periodic_03`.

Finally, I tested hanging nodes over periodic boundaries and they seem fine. I did not try to run a full MG solver on that part so I cannot say 100% sure it is comprehensive for that case, but the periodic connections look good (`mg_ghost_dofs_periodic_04`).

tjhei commented on include/deal.II/distributed/tria.h:
```diff
@@ -966,6 +975,28 @@ namespace parallel
+      /**
+       * This method returns a bit vector of length tria.n_vertices() with the
+       * locally active vertices on a level, i.e., the vertices touched by the
+       * locally owned cells on that level (possibly including the vertices
```
can you say something like "locally owned level cells for use in geometric multigrid"?

tjhei commented:
> Finally, I tested hanging nodes over periodic boundaries and they seem fine. I did not try to run a full MG solver on that part

Excellent, thanks. No need to test running multigrid here. Patch looks good to me (but see the one comment above).

kronbichler commented:
Thanks for the comments. I have fixed the documentation.
